File: ARRAY/Sorting_Implementation/merge_sort.py
def merge_sort(values):
    if len(values) <= 1:
        return values
    # it can not be odd number here be care full maaz 
    middle = len(values) // 2
    left = merge_sort(values[:middle])
    right = merge_sort(values[middle:])
    return merge(left,right)






def merge(left,right):
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result

File: ARRAY/Sorting_Implementation/merge_sort.js
function mergeSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }
    const divide = Math.floor(arr.length / 2);
    const left = mergeSort(arr.slice(0, divide));
    const right = mergeSort(arr.slice(divide));
    return merge(left, right);
}

function merge(left, right) {
    const result = [];
    let i = 0;
    let j = 0;

    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            result.push(left[i]);
            i++;
        } else {
            result.push(right[j]);
            j++;
        }
    }
    while (i < left.length) {
        result.push(left[i]);
        i++;
    }
    while (j < right.length) {
        result.push(right[j]);
        j++;
    }

    return result;
}

console.log(mergeSort([8, 3, 5, 1]));
// [1, 3, 5, 8]